In [1]:
from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


In [2]:
%cd /content/drive/MyDrive/Colab Notebooks

/content/drive/MyDrive/Colab Notebooks


In [3]:
import pandas as pd
from sklearn.model_selection import train_test_split
import time
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import chi2
from sklearn.feature_selection import RFE
from sklearn.linear_model import LogisticRegression
import pickle
import matplotlib.pyplot as plt

In [4]:
def split_scalar(indep_X,dep_Y):
        X_train, X_test, y_train, y_test = train_test_split(indep_X, dep_Y, test_size = 0.25, random_state = 0)
        #X_train, X_test, y_train, y_test = train_test_split(indep_X,dep_Y, test_size = 0.25, random_state = 0)

        #Feature Scaling
        #from sklearn.preprocessing import StandardScaler
        sc = StandardScaler()
        X_train = sc.fit_transform(X_train)
        X_test = sc.transform(X_test)
        return X_train, X_test, y_train, y_test

In [6]:
def r2_prediction(regressor,X_test,y_test):
     y_pred = regressor.predict(X_test)
     from sklearn.metrics import r2_score
     r2=r2_score(y_test,y_pred)
     return r2

def Linear(X_train,y_train,X_test):
        # Fitting K-NN to the Training set
        from sklearn.linear_model import LinearRegression
        regressor = LinearRegression()
        regressor.fit(X_train, y_train)
        r2=r2_prediction(regressor,X_test,y_test)
        return  r2

def svm_linear(X_train,y_train,X_test):

        from sklearn.svm import SVR
        regressor = SVR(kernel = 'linear')
        regressor.fit(X_train, y_train)
        r2=r2_prediction(regressor,X_test,y_test)
        return  r2

def svm_NL(X_train,y_train,X_test):

        from sklearn.svm import SVR
        regressor = SVR(kernel = 'rbf')
        regressor.fit(X_train, y_train)
        r2=r2_prediction(regressor,X_test,y_test)
        return  r2


In [19]:
def Decision(X_train,y_train,X_test):

        # Fitting K-NN to the Training setC
        from sklearn.tree import DecisionTreeRegressor
        regressor = DecisionTreeRegressor(random_state = 0)
        regressor.fit(X_train, y_train)
        r2=r2_prediction(regressor,X_test,y_test)
        return  r2


def random(X_train,y_train,X_test):
        # Fitting K-NN to the Training set
        from sklearn.ensemble import RandomForestRegressor
        regressor = RandomForestRegressor(n_estimators = 10, random_state = 0)
        regressor.fit(X_train, y_train)
        r2=r2_prediction(regressor,X_test,y_test)
        return  r2

def rfeFeature(indep_X,dep_Y,n):
        rfelist=[]
        rfemodellist=[lin,SVRl,dec,rf]
        for i in   rfemodellist:
            print(i)
            log_rfe = RFE(estimator=i, n_features_to_select=n)
            log_fit = log_rfe.fit(indep_X, dep_Y)
            log_rfe_feature=log_fit.transform(indep_X)
            rfelist.append(log_rfe_feature)
        return rfelist

In [9]:
from sklearn.linear_model import LinearRegression
lin = LinearRegression()

from sklearn.svm import SVR
SVRl = SVR(kernel = 'linear')

from sklearn.svm import SVR
#SVRnl = SVR(kernel = 'rbf')

from sklearn.tree import DecisionTreeRegressor
dec = DecisionTreeRegressor(random_state = 0)

from sklearn.ensemble import RandomForestRegressor
rf = RandomForestRegressor(n_estimators = 10, random_state = 0)

In [12]:
def rfe_regression(acclog,accsvml,accdes,accrf):

    rfedataframe=pd.DataFrame(index=['Linear','SVC','Random','DecisionTree'],columns=['Linear','SVMl',
                                                                                        'Decision','Random'])

    for number,idex in enumerate(rfedataframe.index):

        rfedataframe['Linear'][idex]=acclog[number]
        rfedataframe['SVMl'][idex]=accsvml[number]
        rfedataframe['Decision'][idex]=accdes[number]
        rfedataframe['Random'][idex]=accrf[number]
    return rfedataframe




In [16]:
dataset1=pd.read_csv("prep.csv",index_col=None)
df2=dataset1
df2 = pd.get_dummies(df2, drop_first=True)

indep_X=df2.drop('classification_yes', axis=1)
dep_Y=df2['classification_yes']

In [23]:
# =========================
# IMPORT LIBRARIES
# =========================
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.feature_selection import RFE
from sklearn.metrics import r2_score

# =========================
# TRAIN TEST SPLIT + SCALING
# =========================
def split_scalar(indep_X, dep_Y):
    X_train, X_test, y_train, y_test = train_test_split(
        indep_X, dep_Y, test_size=0.25, random_state=0
    )

    sc = StandardScaler()
    X_train = sc.fit_transform(X_train)
    X_test = sc.transform(X_test)

    return X_train, X_test, y_train, y_test


# =========================
# R2 SCORE FUNCTION
# =========================
def r2_prediction(regressor, X_test, y_test):
    y_pred = regressor.predict(X_test)
    r2 = r2_score(y_test, y_pred)
    return r2


# =========================
# LINEAR REGRESSION
# =========================
def Linear(X_train, y_train, X_test, y_test):
    from sklearn.linear_model import LinearRegression
    regressor = LinearRegression()
    regressor.fit(X_train, y_train)
    r2 = r2_prediction(regressor, X_test, y_test)
    return r2


# =========================
# SVM LINEAR
# =========================
def svm_linear(X_train, y_train, X_test, y_test):
    from sklearn.svm import SVR
    regressor = SVR(kernel='linear')
    regressor.fit(X_train, y_train)
    r2 = r2_prediction(regressor, X_test, y_test)
    return r2


# =========================
# SVM NON-LINEAR (RBF)
# =========================
def svm_NL(X_train, y_train, X_test, y_test):
    from sklearn.svm import SVR
    regressor = SVR(kernel='rbf')
    regressor.fit(X_train, y_train)
    r2 = r2_prediction(regressor, X_test, y_test)
    return r2


# =========================
# DECISION TREE REGRESSOR
# =========================
def Decision(X_train, y_train, X_test, y_test):
    from sklearn.tree import DecisionTreeRegressor
    regressor = DecisionTreeRegressor(random_state=0)
    regressor.fit(X_train, y_train)
    r2 = r2_prediction(regressor, X_test, y_test)
    return r2


# =========================
# RANDOM FOREST REGRESSOR
# =========================
def random(X_train, y_train, X_test, y_test):
    from sklearn.ensemble import RandomForestRegressor
    regressor = RandomForestRegressor(n_estimators=10, random_state=0)
    regressor.fit(X_train, y_train)
    r2 = r2_prediction(regressor, X_test, y_test)
    return r2


# =========================
# RFE FEATURE SELECTION
# =========================
def rfeFeature(indep_X, dep_Y, n):
    rfelist = []

    from sklearn.linear_model import LinearRegression
    lin = LinearRegression()

    from sklearn.svm import SVR
    SVRl = SVR(kernel='linear')

    from sklearn.tree import DecisionTreeRegressor
    dec = DecisionTreeRegressor(random_state=0)

    from sklearn.ensemble import RandomForestRegressor
    rf = RandomForestRegressor(n_estimators=10, random_state=0)

    rfemodellist = [lin, SVRl, dec, rf]

    for model in rfemodellist:
        print("RFE Model:", model)
        selector = RFE(estimator=model, n_features_to_select=n)
        fit_model = selector.fit(indep_X, dep_Y)
        selected_features = fit_model.transform(indep_X)
        rfelist.append(selected_features)

    return rfelist


# =========================
# RESULT DATAFRAME
# =========================
def rfe_regression(acclin, accsvml, accdes, accrf):
    rfedataframe = pd.DataFrame(
        index=['Linear', 'SVM_Linear', 'DecisionTree', 'RandomForest'],
        columns=['Linear', 'SVM_Linear', 'DecisionTree', 'RandomForest']
    )

    for number, idx in enumerate(rfedataframe.index):
        rfedataframe.loc[idx, 'Linear'] = acclin[number]
        rfedataframe.loc[idx, 'SVM_Linear'] = accsvml[number]
        rfedataframe.loc[idx, 'DecisionTree'] = accdes[number]
        rfedataframe.loc[idx, 'RandomForest'] = accrf[number]

    return rfedataframe


# =========================
# LOAD DATASET
# =========================
dataset1 = pd.read_csv("prep.csv")   # Upload prep.csv in Colab first

df2 = dataset1.copy()
df2 = pd.get_dummies(df2, drop_first=True)

# Target column
indep_X = df2.drop('classification_yes', axis=1)
dep_Y = df2['classification_yes']


# =========================
# APPLY RFE
# =========================
rfelist = rfeFeature(indep_X, dep_Y, 3)

acclin = []
accsvml = []
accsvmnl = []
accdes = []
accrf = []

for i in rfelist:
    X_train, X_test, y_train, y_test = split_scalar(i, dep_Y)

    r2_lin = Linear(X_train, y_train, X_test, y_test)
    acclin.append(r2_lin)

    r2_sl = svm_linear(X_train, y_train, X_test, y_test)
    accsvml.append(r2_sl)

    r2_NL = svm_NL(X_train, y_train, X_test, y_test)
    accsvmnl.append(r2_NL)

    r2_d = Decision(X_train, y_train, X_test, y_test)
    accdes.append(r2_d)

    r2_r = random(X_train, y_train, X_test, y_test)
    accrf.append(r2_r)


# =========================
# FINAL RESULT
# =========================
result = rfe_regression(acclin, accsvml, accdes, accrf)

print("\nRFE Regression Result:\n")
print(result)

RFE Model: LinearRegression()
RFE Model: SVR(kernel='linear')
RFE Model: DecisionTreeRegressor(random_state=0)
RFE Model: RandomForestRegressor(n_estimators=10, random_state=0)

RFE Regression Result:

                Linear SVM_Linear DecisionTree RandomForest
Linear        0.441961   0.262153     0.441961     0.441816
SVM_Linear    0.441961   0.262153     0.441961     0.441816
DecisionTree  0.664893   0.609652     0.965961     0.916304
RandomForest  0.676174   0.670691     0.933504     0.887256


**After applying RFE feature selection and testing different regression models, the best result was obtained by:

RFE using Decision Tree Regressor

Prediction using Decision Tree Regressor

The highest R² score is 0.965961.

So, Decision Tree Regressor is the best model for this dataset because it gives the highest performance compared to other models.

From the results, it is also clear that Decision Tree and Random Forest perform better than Linear Regression and SVM, which means the dataset has non-linear patterns.

Therefore, the final selected model is Decision Tree Regressor**

In [24]:
# =========================
# APPLY RFE
# =========================
rfelist = rfeFeature(indep_X, dep_Y, 4)

acclin = []
accsvml = []
accsvmnl = []
accdes = []
accrf = []

for i in rfelist:
    X_train, X_test, y_train, y_test = split_scalar(i, dep_Y)

    r2_lin = Linear(X_train, y_train, X_test, y_test)
    acclin.append(r2_lin)

    r2_sl = svm_linear(X_train, y_train, X_test, y_test)
    accsvml.append(r2_sl)

    r2_NL = svm_NL(X_train, y_train, X_test, y_test)
    accsvmnl.append(r2_NL)

    r2_d = Decision(X_train, y_train, X_test, y_test)
    accdes.append(r2_d)

    r2_r = random(X_train, y_train, X_test, y_test)
    accrf.append(r2_r)


# =========================
# FINAL RESULT
# =========================
result = rfe_regression(acclin, accsvml, accdes, accrf)

print("\nRFE Regression Result:\n")
print(result)

RFE Model: LinearRegression()
RFE Model: SVR(kernel='linear')
RFE Model: DecisionTreeRegressor(random_state=0)
RFE Model: RandomForestRegressor(n_estimators=10, random_state=0)

RFE Regression Result:

                Linear SVM_Linear DecisionTree RandomForest
Linear         0.60401   0.457046     0.776711     0.776492
SVM_Linear     0.60401   0.457046     0.776711     0.776492
DecisionTree  0.671727   0.628963     0.835247       0.8403
RandomForest  0.681563   0.614992      0.96711     0.923559


**The best RFE Regression result is 0.96711.**
This is obtained by using Random Forest Regressor for RFE and Decision Tree Regressor for prediction.

So, **Decision Tree Regressor** is selected as the best final model for this dataset.

In [25]:
# =========================
# APPLY RFE
# =========================
rfelist = rfeFeature(indep_X, dep_Y, 5)

acclin = []
accsvml = []
accsvmnl = []
accdes = []
accrf = []

for i in rfelist:
    X_train, X_test, y_train, y_test = split_scalar(i, dep_Y)

    r2_lin = Linear(X_train, y_train, X_test, y_test)
    acclin.append(r2_lin)

    r2_sl = svm_linear(X_train, y_train, X_test, y_test)
    accsvml.append(r2_sl)

    r2_NL = svm_NL(X_train, y_train, X_test, y_test)
    accsvmnl.append(r2_NL)

    r2_d = Decision(X_train, y_train, X_test, y_test)
    accdes.append(r2_d)

    r2_r = random(X_train, y_train, X_test, y_test)
    accrf.append(r2_r)


# =========================
# FINAL RESULT
# =========================
result = rfe_regression(acclin, accsvml, accdes, accrf)

print("\nRFE Regression Result:\n")
print(result)

RFE Model: LinearRegression()
RFE Model: SVR(kernel='linear')
RFE Model: DecisionTreeRegressor(random_state=0)
RFE Model: RandomForestRegressor(n_estimators=10, random_state=0)

RFE Regression Result:

                Linear SVM_Linear DecisionTree RandomForest
Linear        0.620124   0.457136      0.77924     0.780135
SVM_Linear    0.604508   0.456871     0.776474     0.776745
DecisionTree  0.674403   0.628206     0.696181     0.815538
RandomForest  0.686361   0.643365     0.836806     0.845303


The best **RFE Regression result is 0.845303.**
This is obtained by using Random Forest Regressor for RFE and Random Forest Regressor for prediction.

So, **Random Forest Regressor** is selected as the best final model for this dataset.

**Row = Feature selection**,
**Column = Prediction**,
**Highest score = Final best combination**

#Best RFE selector = RandomForest

#Best final prediction model = DecisionTree